Check matched search text for word breaks

keewhip opened the pull request:
Adjust starting position and length of the returned text when a search match is found and the length of the searched text is changed by SearchHelper::remove_accents
This also removes the trailing ... when the searched text is at it's end.
 #30348

Pull Request for Issue # .

### Summary of Changes



### Testing Instructions



### Actual result BEFORE applying this Pull Request



### Expected result AFTER applying this Pull Request



### Documentation Changes Required

keewhip committed:
Check matched search text for word breaks

Adjust starting position and length of the returned text when a search match is found and the length of the searched text is changed by SearchHelper::remove_accents
This also removes the trailing ... when the searched text is at it's end.
 #30348

Quy commented on administrator/components/com_search/helpers/search.php:
```diff
@@ -310,7 +310,33 @@ public static function _smartSubstr($text, $searchword)
-			return (($pos > 0) ? '...&#160;' : '') . StringHelper::substr($text, $pos, $chunk_size) . '&#160;...';
+			// Check if original text is other length than searched text (changed by self::remove_accents)
+			$iOriLen = StringHelper::strlen(StringHelper::substr($text, 0, $pos + $chunk_size));
+			$iModLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, 0, $pos + $chunk_size)));
+			if($iOriLen != $iModLen)
+			{
+				//check if difference is in displayed text or before
+				if ($pos == 0) // displayed text only, adjust $chunk_size
+				{
+					$chunk_size += $iOriLen - $iModLen;
+				}
+				else
+				{
+					$iOriSkippedLen = StringHelper::strlen(StringHelper::substr($text, 0, $pos));
+					$iModSkippedLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, 0, $pos)));
+					if ($iOriSkippedLen != $iModSkippedLen) // adjust starting position $pos
+					{
+						$pos += $iOriSkippedLen - $iModSkippedLen;
+					}
+					$iOriReturnLen = StringHelper::strlen(StringHelper::substr($text, $pos, $chunk_size));
+					$iModReturnLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, $pos, $chunk_size)));
+					if ($iOriReturnLen != $iModReturnLen)
+					{
+						$chunk_size += $iOriReturnLen - $iModReturnLen;
+					}
+				}
+			}
+			return (($pos > 0) ? '...&#160;' : '') . StringHelper::substr($text, $pos, $chunk_size) . ($pos + $chunk_size >= StringHelper::strlen($text) ? '' : '&#160;...');
```
```suggestion
			$iModLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, 0, $pos + $chunk_size)));

			if ($iOriLen != $iModLen)
			{
				//check if difference is in displayed text or before
				if ($pos == 0) // displayed text only, adjust $chunk_size
				{
					$chunk_size += $iOriLen - $iModLen;
				}
				else
				{
					$iOriSkippedLen = StringHelper::strlen(StringHelper::substr($text, 0, $pos));
					$iModSkippedLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, 0, $pos)));

					if ($iOriSkippedLen != $iModSkippedLen) // adjust starting position $pos
					{
						$pos += $iOriSkippedLen - $iModSkippedLen;
					}

					$iOriReturnLen = StringHelper::strlen(StringHelper::substr($text, $pos, $chunk_size));
					$iModReturnLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, $pos, $chunk_size)));

					if ($iOriReturnLen != $iModReturnLen)
					{
						$chunk_size += $iOriReturnLen - $iModReturnLen;
					}
				}
			}

			return (($pos > 0) ? '...&#160;' : '') . StringHelper::substr($text, $pos, $chunk_size) . ($pos + $chunk_size >= StringHelper::strlen($text) ? '' : '&#160;...');
```

Commit:
Update administrator/components/com_search/helpers/search.php

coding style

keewhip committed:
Always check the positions and sizes of the search chunks

Always perform the check for the right staring postion and the correct chunk size with the original text and the modified search text (with remove_accents)
Size can be the same for the total text but not for certain parts

keewhip committed:
Fix Line exceeds 150 characters error

keewhip committed:
coding style

keewhip committed:
force trigger check builds

SharkyKZ commented on administrator/components/com_search/helpers/search.php:
```diff
@@ -310,7 +310,36 @@ public static function _smartSubstr($text, $searchword)
+				$iOriSkippedLen = StringHelper::strlen(StringHelper::substr($text, 0, $pos));
+				$iModSkippedLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, 0, $pos)));
+				
+				if ($iOriSkippedLen != $iModSkippedLen) // Adjust starting position $pos
```
```suggestion
				// Adjust starting position $pos
				if ($iOriSkippedLen != $iModSkippedLen)
```

keewhip committed:
code style

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

SharkyKZ commented on administrator/components/com_search/helpers/search.php:
```diff
@@ -310,7 +310,37 @@ public static function _smartSubstr($text, $searchword)
 		{
-			return (($pos > 0) ? '...&#160;' : '') . StringHelper::substr($text, $pos, $chunk_size) . '&#160;...';
+			// Check if original text is different length than searched text (changed by function self::remove_accents)
+			if ($pos == 0) // Displayed text only, adjust $chunk_size
```
```suggestion
			// Displayed text only, adjust $chunk_size
			if ($pos === 0)
```

SharkyKZ commented on administrator/components/com_search/helpers/search.php:
```diff
@@ -310,7 +310,37 @@ public static function _smartSubstr($text, $searchword)
+				$iModSkippedLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, 0, $pos)));
+				
+				// Adjust starting position $pos
+				if ($iOriSkippedLen != $iModSkippedLen)
```
```suggestion
				if ($iOriSkippedLen !== $iModSkippedLen)
```

SharkyKZ commented on administrator/components/com_search/helpers/search.php:
```diff
@@ -310,7 +310,37 @@ public static function _smartSubstr($text, $searchword)
+				$iOriReturnLen = StringHelper::strlen(StringHelper::substr($text, $pos, $chunk_size));
+				$iModReturnLen = StringHelper::strlen(self::remove_accents(StringHelper::substr($text, $pos, $chunk_size)));
+				
+				if ($iOriReturnLen != $iModReturnLen)
```
```suggestion
				if ($iOriReturnLen !== $iModReturnLen)
```

keewhip committed:
Strict comparison for $pos variable

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

SharkyKZ commented on administrator/components/com_search/helpers/search.php:
```diff
@@ -310,7 +310,38 @@ public static function _smartSubstr($text, $searchword)
+			
+			$sPre = $pos > 0 ? '...&#160;' : '';
+			$sPost = ($pos + $chunk_size) >= StringHelper::strlen($text) ? '' : '&#160;...';
+			return $sPre . StringHelper::substr($text, $pos, $chunk_size) . $sPost;
```
```suggestion

			return $sPre . StringHelper::substr($text, $pos, $chunk_size) . $sPost;
```

Formatio-hippocampi commented:
I have tested this item :white_check_mark: successfully on d4b99f1f921c147c843aaeab5b057dec20cc285f<br /><br />See what to test at #30348<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30350">issues.joomla.org/tracker/joomla-cms/30350</a>.</sub>

Commit:
Update administrator/components/com_search/helpers/search.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

SharkyKZ commented:
I have tested this item :white_check_mark: successfully on af12d2ea273a239d7915bb7f218d9e7b1407b170<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30350">issues.joomla.org/tracker/joomla-cms/30350</a>.</sub>

SharkyKZ commented:
RTC.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30350">issues.joomla.org/tracker/joomla-cms/30350</a>.</sub>
